mm2src/mm2_main/src/rpc/lp_commands/lr_swap/lr_impl.rs

```diff
@@ -0,0 +1,350 @@
+                ((src, dst), dst_price)
+            })
+            .collect::<HashMap<_, _>>();
+        for q in &quotes {
+            log::debug!(
+                "query_destination_token_prices src/dst quote={:?} {}",
+                q,
+                q.1.to_decimal()
+            );
+        }
+        self.update_with_lr_prices(quotes);
+        Ok(())
+    }
+
+    /// Estimate the needed source amount for LR swap, by dividing the known dst amount by the src/dst price
+    #[allow(clippy::result_large_err)]
+    fn estimate_source_token_amounts(&mut self) -> MmResult<(), ApiIntegrationRpcError> {
+        for lr_data in self.inner.values_mut() {
+            let Some(ref dst_price) = lr_data.lr_price else {
+                continue;
+            };
+            let dst_amount = lr_data
+                .dst_amount
+                .ok_or(ApiIntegrationRpcError::InternalError("no dst_amount".to_owned()))?;
+            let dst_amount = mm_number_from_u256(dst_amount);
+            if let Some(src_amount) = &dst_amount.checked_div(dst_price) {
+                lr_data.src_amount = Some(mm_number_to_u256(src_amount)?);
+                log::debug!(
+                    "estimate_source_token_amounts lr_data.order.coin={} dst_price={} lr_data.src_amount={:?}",
+                    lr_data.order.coin,
+                    dst_price.to_decimal(),
+                    lr_data.src_amount
+                );
+            }
+        }
+        Ok(())
+    }
+
+    /// Run 1inch requests to get LR quotes to convert source tokens to tokens in orders
+    async fn run_lr_quotes(&mut self, ctx: &MmArc) -> MmResult<(), ApiIntegrationRpcError> {
+        let mut src_dst = vec![];
+        let mut quote_futs = vec![];
+        for ((src_token, dst_token), lr_data) in self.inner.iter() {
+            let Some(src_amount) = lr_data.src_amount else {
+                continue;
+            };
+            let (src_contract, dst_contract, chain_id) = lr_data.get_chain_contract_info()?;
+            let query_params = ClassicSwapQuoteParams::new(src_contract, dst_contract, src_amount.to_string())
+                .with_include_tokens_info(Some(true))
+                .with_include_gas(Some(true))
+                .build_query_params()?;
+            let url_builder = SwapUrlBuilder::create_api_url_builder(ctx, chain_id, SwapApiMethods::ClassicSwapQuote)?;
+            let fut = ApiClient::call_one_inch_api::<ClassicSwapData>(url_builder, Some(query_params));
+            quote_futs.push(fut);
+            src_dst.push((src_token.clone(), dst_token.clone()));
+        }
+
+        let swap_data = join_all(quote_futs)
+            .await
+            .into_iter()
+            .filter_map(|res| if let Ok(lr_data) = res { Some(lr_data) } else { None }) // skip if bad result received (for e.g. low liguidity)
+            .collect::<Vec<_>>();
+        let swap_data_map = src_dst.into_iter().zip(swap_data.into_iter()).collect();
```

Comment:
when filtering out errors then zipping with the src_dst pairs, the result might mis-align if some errors were filtered out.

Reply:
fixed thanks